pkl-core/src/main/java/org/pkl/core/runtime/TextFormatter.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core.runtime;
+
+import java.io.PrintWriter;
+import org.pkl.core.util.StringBuilderWriter;
+
+/*
+   TODO:
+     * Make "margin matter" a facility of the formatter, managing margins in e.g. `newline()`.
+      - `pushMargin(String matter)` / `popMargin()`
+     * Replace implementation methods `repeat()` and `repeatError()` with more semantic equivalents.
+      - `underline(int startColumn, int endColumn)`
+     * Replace `newInstance()` with an alternative that doesn't require instance management,
+       i.e. better composition (currently only used for pre-rendering `hint`s).
+     * Assert assumed invariants (e.g. `text(String text)` checking there are no newlines).
+*/
+public abstract class TextFormatter<SELF extends TextFormatter<SELF>> {
+  public static TextFormatter<?> create(boolean usingColor) {
+    return usingColor ? new AnsiFormatter() : new PlainFormatter();
+  }
+
+  protected final StringBuilder builder = new StringBuilder();
+
+  public PrintWriter toPrintWriter() {
+    return new PrintWriter(new StringBuilderWriter(builder));
+  }
+
+  public String toString() {
+    return builder.toString();
+  }
+
+  public SELF newline() {
+    return newlines(1);
+  }
+
+  public abstract SELF newInstance();
+
+  public abstract SELF newlines(int count);
+
+  // ---- Styling methods:
+
+  public abstract SELF margin(String marginMatter);
+
+  public abstract SELF hint(String hint);
+
+  public abstract SELF stackOverflowLoopCount(int counter);
+
+  public abstract SELF lineNumber(String line);
+
+  public abstract SELF text(String text);
+
+  public abstract SELF object(Object obj);
+
+  public abstract SELF errorHeader(String header);
+
+  public abstract SELF error(String message);
```

Comment:
General note: I think a better API is to just have methods that describe _what_ the format is. For example:

```java
out
  .bold()
  .color(Color.RED)
  .append("–– Pkl Error ––")
  .append("\n")
```

This feels much easier to use when we want to extend colors to other places.

If the goal is to be able to theme this later on, that can happen in when these messages are built.

```java
out
  .bold()
  .color(errorHeaderColor)
  .append("–– Pkl Error ––")
  .reset()
  .append("\n")
```

Here, `errorHeaderColor` can come from some config file or something.

Also: I don't know if we really need multiple subclasses; it seems good enough to just have a text formatter that accepts `boolean color` as a constructor arg.

Given this is all private, we don't need to figure this out right now. But, I'm tempted to say that we should refactor this later.

Reply:
If colors are disabled, `.color(Color.RED)` would just be a no-op!